tests/libvmi/vmi.go

```diff
@@ -0,0 +1,151 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * Copyright 2020 Red Hat, Inc.
+ *
+ */
+
+package libvmi
+
+import (
+	k8sv1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+	k8smetav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/rand"
+
+	kvirtv1 "kubevirt.io/client-go/api/v1"
+)
+
+type Option func(vmiOptions)
+
+const DefaultVmiName = "testvmi"
+
+type action func(vmi *kvirtv1.VirtualMachineInstance)
+type vmiOptions map[int]action
+
+const defaultTestGracePeriod int64 = 0
+
+const (
+	// Options Identifiers
+	optNone = iota
+	optEnableRandomName
+
+	optContainerImageName
+	optEnableRng
+	optResourceMemory
+	optInterface
+	optNetwork
+	optTerminationGracePeriod
+
+	optCloudInitNoCloudUserData
+	optCloudInitNoCloudUserDataB64
+	optCloudInitNoCloudNetworkData
+	optCloudInitNoCloudNetworkDataB64
+)
+
+var defaultVmiOptions = vmiOptions{
+	optEnableRandomName: nil,
+	optInterface: func(vmi *kvirtv1.VirtualMachineInstance) {
+		vmi.Spec.Domain.Devices.Interfaces = append(
+			vmi.Spec.Domain.Devices.Interfaces, primaryInterfaceDevice(),
+		)
+	},
+	optNetwork: func(vmi *kvirtv1.VirtualMachineInstance) {
+		vmi.Spec.Networks = append(vmi.Spec.Networks, *kvirtv1.DefaultPodNetwork())
+	},
+	optTerminationGracePeriod: func(vmi *kvirtv1.VirtualMachineInstance) {
+		t := defaultTestGracePeriod
+		vmi.Spec.TerminationGracePeriodSeconds = &t
+	},
+	optResourceMemory: func(vmi *kvirtv1.VirtualMachineInstance) {
+		vmi.Spec.Domain.Resources.Requests = k8sv1.ResourceList{
+			k8sv1.ResourceMemory: resource.MustParse("8192Ki"),
+		}
+	},
+}
+
+// WithRandomName adds a random text to the vmi name.
+func WithRandomName(enableRandomName bool) Option {
+	return func(o vmiOptions) {
+		if enableRandomName {
+			o[optEnableRandomName] = nil
+		} else {
+			delete(o, optEnableRandomName)
+		}
+	}
+}
+
+// WithRng adds `rng` to the the vmi devices.
+func WithRng() Option {
+	return func(o vmiOptions) {
+		o[optEnableRng] = func(vmi *kvirtv1.VirtualMachineInstance) {
+			vmi.Spec.Domain.Devices.Rng = &kvirtv1.Rng{}
+		}
+	}
+}
+
+// WithResourceMemory specifies the vmi memory resource.
+func WithResourceMemory(value string) Option {
+	return func(o vmiOptions) {
+		o[optResourceMemory] = func(vmi *kvirtv1.VirtualMachineInstance) {
+			vmi.Spec.Domain.Resources.Requests = k8sv1.ResourceList{
+				k8sv1.ResourceMemory: resource.MustParse(value),
+			}
+		}
+	}
+}
+
+// New instantiates a new VMI configuration,
+// building its properties based on the specified With* options.
+func New(namespace, name string, opts ...Option) *kvirtv1.VirtualMachineInstance {
+	options := defaultVmiOptions
+	for _, opt := range opts {
+		opt(options)
+	}
+
+	// Randomizing the VMI name is a special case which must occur before
+	// creating the VMI.
+	if _, ok := options[optEnableRandomName]; ok {
```

Comment:
Why don't we implement this as an normal option at default ones ? `WIthNameSuffix` at default, so is all at the same place.

Reply:
We need this examined before the vmi object is created, because it is not that straight forward to update a given name.